scipy/interpolate/_rgi_cython.pyx

```diff
@@ -0,0 +1,164 @@
+
+    """
+    cdef int interval, high, low, mid
+    cdef double a, b
+
+    a = x[0]
+    b = x[nx-1]
+
+    interval = prev_interval
+    if interval < 0 or interval >= nx:
+        interval = 0
+
+    if not (a <= xval <= b):
+        # Out-of-bounds (or nan)
+        if xval < a and extrapolate:
+            # below
+            interval = 0
+        elif xval > b and extrapolate:
+            # above
+            interval = nx - 2
+        else:
+            # nan or no extrapolation
+            interval = -1
+    elif xval == b:
+        # Make the interval closed from the right
+        interval = nx - 2
+    else:
+        # Find the interval the coordinate is in
+        # (binary search with locality)
+        if xval >= x[interval]:
+            low = interval
+            high = nx - 2
+        else:
+            low = 0
+            high = interval
+
+        if xval < x[low+1]:
+            high = low
+
+        while low < high:
+            mid = (high + low)//2
+            if xval < x[mid]:
+                # mid < high
+                high = mid
+            elif xval >= x[mid + 1]:
+                low = mid + 1
+            else:
+                # x[mid] <= xval < x[mid+1]
+                low = mid
+                break
+
+        interval = low
+
+    return interval
+
+
+def evaluate_linear(values, indices, norm_distances, out_of_bounds):
+    # slice for broadcasting over trailing dimensions in self.values
+    vslice = (slice(None),) + (None,) * (values.ndim - len(indices))
+
+    # Compute shifting up front before zipping everything together
+    shift_norm_distances = [1 - yi for yi in norm_distances]
+    shift_indices = [i + 1 for i in indices]
+
+    # The formula for linear interpolation in 2d takes the form:
+    # values = self.values[(i0, i1)] * (1 - y0) * (1 - y1) + \
+    #          self.values[(i0, i1 + 1)] * (1 - y0) * y1 + \
+    #          self.values[(i0 + 1, i1)] * y0 * (1 - y1) + \
+    #          self.values[(i0 + 1, i1 + 1)] * y0 * y1
+    # We pair i with 1 - yi (zipped1) and i + 1 with yi (zipped2)
+    zipped1 = zip(indices, shift_norm_distances)
+    zipped2 = zip(shift_indices, norm_distances)
+
+    # Take all products of zipped1 and zipped2 and iterate over them
+    # to get the terms in the above formula. This corresponds to iterating
+    # over the vertices of a hypercube.
+    hypercube = itertools.product(*zip(zipped1, zipped2))
+    value = np.array([0.])
+    for h in hypercube:
+        edge_indices, weights = zip(*h)
+        weight = np.array([1.])
+        for w in weights:
+            weight = weight * w
+        value = value + np.asarray(values[edge_indices]) * weight[vslice]
+    return value
+
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+@cython.cdivision(True)
+def find_indices(grid, double[:, :] xi):
+    cdef long i, j, I, J, index, grid_i_size, nx
+    cdef double denom, grid_i_index
+    cdef np.ndarray[long, ndim=2] indices
```

Comment:
Can this be a long[:, ::1]

Reply:
I don't think so, unless we convert it back to an ndarray afterwards. Here's the exception I get when I tried changing it:

```python
File "_rgi_cython.pyx", line 101, in scipy.interpolate._rgi_cython.evaluate_linear
               TypeError: unsupported operand type(s) for +: 'scipy.interpolate._rgi_cython._memoryviewslice' and 'int'
```

Reply:
Exactly, take the buffer before looping, index memoryviews, do asarray(...) on return. This may simplify indexing in hot loops. Or not --- needs checking.